azurelinuxagent/ga/remoteaccess.py

```diff
@@ -48,108 +47,107 @@ def get_remote_access_handler(protocol):
+            # were never added. Enumerating the users on the current vm can be very slow and this path is hit
+            # on each goal state; we use self._check_existing_jit_users to avoid enumerating the users
+            # every single time.
+            if self._check_existing_jit_users:
+                logger.info("Looking for existing remote access users.")
+
+                self._check_existing_jit_users = False
+
+                existing_jit_users = self._get_existing_jit_users()
+
+                for user in existing_jit_users:
+                    try:
+                        logger.info("Removing remote access user '{0}'", user)
+                        self._remove_user(user)
+                    except Exception as e:
+                        logger.error("Error removing remote access user '{0}' - {1}", user, ustr(e))
+
+    @staticmethod
+    def _is_jit_user(comment):
         return comment == REMOTE_ACCESS_ACCOUNT_COMMENT
 
-    def add_user(self, username, encrypted_password, account_expiration):
+    def _add_user(self, username, encrypted_password, account_expiration):
+        user_added = False
+
         try:
             expiration_date = (account_expiration + timedelta(days=1)).strftime(DATE_FORMAT)
-            logger.verbose("Adding user {0} with expiration date {1}".format(username, expiration_date))
-            self.os_util.useradd(username, expiration_date, REMOTE_ACCESS_ACCOUNT_COMMENT)
-        except Exception as e:
-            raise RemoteAccessError("Error adding user {0}. {1}".format(username, ustr(e)))
-        try:
-            prv_key = os.path.join(conf.get_lib_dir(), TRANSPORT_PRIVATE_CERT)
-            pwd = self.cryptUtil.decrypt_secret(encrypted_password, prv_key)
-            self.os_util.chpasswd(username, pwd, conf.get_password_cryptid(), conf.get_password_crypt_salt_len())
-            self.os_util.conf_sudoer(username)
-            logger.info("User '{0}' added successfully with expiration in {1}".format(username, expiration_date))
-        except Exception as e:
-            error = "Error adding user {0}. {1} ".format(username, str(e))
-            try:
-                self.handle_failed_create(username)
-                error += "cleanup successful"
-            except RemoteAccessError as rae:
-                error += "and error cleaning up {0}".format(str(rae))
-            raise RemoteAccessError("Error adding user {0} cleanup successful".format(username), ustr(e))
-
-    def handle_failed_create(self, username):
-        try:
-            self.delete_user(username)
-        except Exception as e:
-            raise RemoteAccessError("Failed to clean up after account creation for {0}.".format(username), e)
+            logger.info("Adding remote access user '{0}' with expiration date {1}", username, expiration_date)
+            self._os_util.useradd(username, expiration_date, REMOTE_ACCESS_ACCOUNT_COMMENT)
```

Comment:
Wouldn't it be better to move this line to outside of the try catch? We don't actually have any handling logic for Exceptions that occur before the user_added flag is set (line 138), so any Exception raised by this line has to be caught by a calling function's try: except anyway. This way we wouldn't have to keep track of the user_added flag at all.

Reply:
I wrote the code like that initially but then I thought it looked more natural if the exception handling covers the entire function and we have a flag for when to clean up.